### DISCUSSION TITLE
AbstractLoginFormAuthenticator::supports() always returns false if application lives under a directory

### DISCUSSION THREAD

**Discussion by php4fan** [symfony/symfony#44329] [Q&A]
### Symfony version(s) affected

5.4.0

### Description

Say you have an application deployed to `https://somedomain.com/my/path`, as opposed to just `https://somedomain.com/`.
I think you usually call that an application that "lives under a subdirectory" (or directory, or folder).

I expect EVERYTHING in the framework to work out of the box with ZERO configuration in that scenario. Generally, Symfony does know how to handle that.

Out-of-the-box login authentication however seems to be one of those things that do not work as expected in the under-a-directory scenario.

I think that's due to the `AbstractLoginFormAuthenticator::supports()` method whose current implementation is:
```
    return $request->isMethod('POST') && $this->getLoginUrl($request) === $request->getPathInfo();
```

In this scenario, `$this->getLoginUrl($request)` returns `/my/path/login`, while `$request->getPathInfo()` returns just `/login` when the url being requested is `/my/path/info`.

### How to reproduce

Deploy a Symfony application under a directory. 
Make a `LoginFormAuthenticator` with `make:auth`
Try to log in.

**Comment by berkut1** [comment]
What about this?
`return 'app_login' === $request->attributes->get('_route') && $request->isMethod('POST');`
We comparing not path, but the name of action.

**Reply by php4fan** [reply]
See @ahundiak 's point about this:

> But don't suggest using _route in the symfony-http package. _route is framework specific functionality.

**Comment by wouterj** [comment]
Fyi, I'll move this one to GitHub discussions.

This does not seem to be an issue with the authenticator interface - but rather a question about the implementation of the method. If, in the end, you come up with a better solution, you can maybe fix the auto generated code in the [Symfony MakerBundle](https://github.com/symfony/maker-bundle). Good luck! :)

**Comment by php4fan** [comment]
No, this is not just about the auto-generated code. 

This is a bug in `AbstractLoginFormAuthenticator` which lives in `symfony/security-http`

**Reply by php4fan** [reply]
@wouterj the issue should be reopened

**Comment by ahundiak** [comment]
For what it is worth there have been a [number of stackoverflow questions](https://stackoverflow.com/questions/68837184/symfony-simple-login-form-with-the-new-authentication-method-not-working#comment121656328_68837184) on this topic.  You setup a simple form authenticator using the default tools and it no workee.  Eventually turns out that the supports method fails.  Various workarounds fix the problem.

Might be worth while to have make:auth simply add a supports method to the generated auth class just to make it more obvious that the method might need to be overridden.

**Reply by php4fan** [reply]
Yep, it's not that it "might need" to be overridden, it is broken and needs to be overridden. It _might work_ out of sheer luck (i.e. if your app happens to live at the root of the domain). If `AbstractLoginFormAuthenticator::supports()` is not guaranteed to work (at least under normal circumstances, outside of absurdly weird or terribly specific scenarios, which the-app-living-in-a-subdirectory is not), then it should be an abstract method so you are _forced_ to override it... But then why not just implement it correctly in the first place?

**Reply by ahundiak** [reply]
Just for info, using phrases like 'absurdly weird' and 'terribly specific' is probably not the best approach to influencing Symfony developers in a positive fashion.  The default supports method works fine for me.  Can't remember the last time I used application subdirectories when deploying an app.  I am not a big fan of the generated authentication class myself so I don't use it.  No biggee.  I don't care for the AbstractController class either.

I'd suggest submitting a PR and then moving on.  But don't suggest using _route in the symfony-http package.  _route is framework specific functionality.

**Comment by weaverryan** [comment]
I've just transferred this back to an issue: https://github.com/symfony/symfony/issues/44893
